engine/src/execution_graph/logic_controllers/BatchWindowFunctionProcessing.cpp

```diff
@@ -32,48 +35,95 @@ ComputeWindowKernel::ComputeWindowKernel(std::size_t kernel_id, const std::strin
-        table_to_rolling.push_back(input_cudf_view.column(this->column_indices_partitioned[col_i]));
+        columns_to_partition.push_back(input_table_cudf_view.column(this->column_indices_partitioned[col_i]));
+        column_partitioned_indices.push_back(col_i);
     }
 
-    cudf::table_view table_view_with_single_col(table_to_rolling);
-
+    cudf::table_view partitioned_table_view(columns_to_partition);
+    
     std::unique_ptr<CudfColumn> windowed_col;
-    if (this->expression.find("ORDER BY") != std::string::npos) {
-        if (this->expression.find("BETWEEN") != std::string::npos) {
-            if (this->expression.find("RANGE") != std::string::npos) {
-                throw std::runtime_error("In Window Function: RANGE is not currently supported");
-            }
-            windowed_col = cudf::grouped_rolling_window(table_view_with_single_col, input_col_view, this->preceding_value + 1, this->following_value, 1, window_aggregation);
+    if (is_first_value_window(this->type_aggs_as_str[pos]) || is_last_value_window(this->type_aggs_as_str[pos])) {
+
+        // first: we want to get all the first (or last) values (due to the columns to partition)
+        std::vector<cudf::groupby::aggregation_request> requests;
+        requests.emplace_back(cudf::groupby::aggregation_request());
+        requests[0].values = col_view_to_agg;
+        requests[0].aggregations.push_back(std::move(window_aggregation));
+        
+        cudf::groupby::groupby gb_obj(cudf::table_view({partitioned_table_view}), cudf::null_policy::EXCLUDE, cudf::sorted::YES, {}, {});
+        std::pair<std::unique_ptr<cudf::table>, std::vector<cudf::groupby::aggregation_result>> result = gb_obj.aggregate(requests);
+
+        windowed_col = std::move(result.second[0].results[0]);
+
+        // if exists duplicated values (in partitioned_table_view) we want to fill `windowed_col` with repeated values
+        // So let's do a left join
+        if (windowed_col->size() < col_view_to_agg.size()) {
+            std::vector<std::unique_ptr<cudf::column>> keys_grouped = result.first->release();
+            keys_grouped.push_back(std::move(windowed_col)); 
+
+            std::unique_ptr<cudf::table> left_table = std::make_unique<cudf::table>(std::move(keys_grouped));
+
+            // Let's get all the necessary params for the left join
+            std::vector<std::pair<cudf::size_type, cudf::size_type>> columns_in_common;
+
+            // we just want the key columns, not the values column (which is the last column)
+            std::vector<cudf::size_type> left_column_indices(left_table->num_columns() - 1);
+            std::iota(left_column_indices.begin(), left_column_indices.end(), 0);
+
+            std::vector<cudf::size_type> right_column_indices(partitioned_table_view.num_columns());
+            std::iota(right_column_indices.begin(), right_column_indices.end(), 0);
+
+            std::unique_ptr<cudf::table> join_table = cudf::left_join(
```

Comment:
why are you doing a left_join as opposed to an inner_join? is it because of nulls?
Both sides of join should have the same keys, so it seems like an inner_join would work. 

Also, if you are doing a left join, I would have expected to have the left side be `partitioned_table_view` and not `left_table`. `left_table` contains the group by keys, which would contain the first or last value per partition. You would want to add the corresponding first or last value to `partitioned_table_view` which has all the values. Honestly, i am not sure how much it matters, except for  perhaps the expected order of the output, and the order matters.